### PR TITLE
Remove mutable default argument

### PR DESCRIPTION
**Pull request by kuenishi**
Also I updated some docs of `FileCache`.

### CODE REVIEW

File: chainerio/cache/file_cache.py [hunk 1]
```diff
@@ -82,16 +82,30 @@ class FileCache(cache.Cache):
     Stores cache data in local filesystem,
     ``~/.chainer/chainerio/cache`` by default. Cache data is
     automatically deleted after the object is collected. If the
-    process quit (typically by SIGTERM) the cache remains after the
+    process quits (typically by SIGTERM) the cache remains after the
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: "If the process quits (typically by SIGTERM)*,* the cache remains after the .."

File: chainerio/cache/file_cache.py [hunk 2]
```diff
@@ -82,16 +82,30 @@ class FileCache(cache.Cache):
...
     death of process.
 
+    Arguments:
+        length (int): Length of the cache array. Typically it is
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: "Typically, ..."

File: chainerio/cache/file_cache.py [hunk 3]
```diff
@@ -82,16 +82,30 @@ class FileCache(cache.Cache):
...
+            the length of Chainer's ``Dataset``.
+        multithread_safe (bool): Defines multithread safety. If this
+            is ``True``, reader-writer locking is introduced behind
+            the cache system. Typically it is for Chainer's
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: "Typically, ..."

Comment (belldandyxtq) [reply chainerio/cache/file_cache.py]: Should we mention the lock overhead here?

File: chainerio/cache/file_cache.py [hunk 4]
```diff
@@ -82,16 +82,30 @@ class FileCache(cache.Cache):
...
+            ``MultithreadIterator``.
+        do_pickle (bool):
+            Do automatic pickle and unpickle inside the cache.
+        dir (str):
+            The place for cache data in case home directory is not
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: The directory where the cache data will be stored. In the default case, the home directory is used. Recommend to fast storage device and avoid using remote filesystems such NFS

File: chainerio/cache/file_cache.py [hunk 5]
```diff
@@ -77,21 +77,33 @@ def unlock(self):
 
 
 class FileCache(cache.Cache):
-    '''Stores cache data in local filesystem
+    '''Cache system with local filesystem
 
-    Stores cache data in local filesystem,
+    Stores cached data in local filesystem,
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: "Cache data on local directory, "
As ``~/.chainer/chainerio/cache`` is not a filesystem

File: chainerio/cache/file_cache.py [hunk 6]
```diff
@@ -77,21 +77,33 @@ def unlock(self):
...
-    process quit (typically by SIGTERM) the cache remains after the
-    death of process.
+    automatically deleted after the object is collected. When this
+    object is not correctly closed, (e.g., the process killed by
+    SIGTERM), the cache remains after the death of process.
+
+    Arguments:
+        length (int): Length of the cache array.
+        multithread_safe (bool): Defines multithread safety. If this
+            is ``True``, reader-writer locking is introduced behind
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: A readers-writer lock is introduced

File: chainerio/cache/file_cache.py [hunk 7]
```diff
@@ -77,21 +77,33 @@ def unlock(self):
...
+            the cache system. Major use case is with Chainer's
+            ``MultithreadIterator``.
+        do_pickle (bool):
+            Do automatic pickle and unpickle inside the cache.
+        dir (str):
+            The path to directory to place cache data in case home
```
Comment (belldandyxtq) [inline chainerio/cache/file_cache.py]: https://github.com/chainer/chainerio/pull/63#discussion_r327915198

File: chainerio/cache/file_cache.py (hunk 6 above)
Comment (belldandyxtq) [reply chainerio/cache/file_cache.py]: https://github.com/chainer/chainerio/pull/63#discussion_r327914793

Comment (kuenishi) [reply chainerio/cache/file_cache.py]: Mentioning lock primitive and design is not only for overhead. 1) When designing a multi-threaded system, the locking primitive behind it is very important. For example in Python, `asyncio.Lock` and `threading.Lock` can't be mixed without *practical* level of carefulness. Also it is not easily safe to mix with `multiprocessing.Lock`. Thus I added  description on what primitive is used.
2) The type of lock is important for app developer, too. Understanding that `get` doesn't block another `get` but `put` blocks `get` is important to design multithreaded and read-intensive program like ours. The design would change a lot if it is just a single mutex, where `get` blocks `get`.

File: chainerio/cache/file_cache.py (hunk 7 above)
Comment (kuenishi) [reply chainerio/cache/file_cache.py]: I didn't get what you meant. Could elaborate a bit more?

File: chainerio/cache/file_cache.py (hunk 6 above)
Comment (kuenishi) [reply chainerio/cache/file_cache.py]: Mentioning lock primitive and design is not only for overhead. 1) When designing a multi-threaded system, the locking primitive behind it is very important. For example in Python, `asyncio.Lock` and `threading.Lock` can't be mixed without *practical* level of carefulness. Also it is not easily safe to mix with `multiprocessing.Lock`. Thus I added  description on what primitive is used.
2) The type of lock is important for app developer, too. Understanding that `get` doesn't block another `get` but `put` blocks `get` is important to design multithreaded and read-intensive program like ours. The design would change a lot if it is just a single mutex, where `get` blocks `get`.

File: chainerio/cache/file_cache.py (hunk 7 above)
Comment (belldandyxtq) [reply chainerio/cache/file_cache.py]: Sorry to make it confusing. It was a suggestion, to change the description of "dir" to

```
The directory where the cache data will be stored. In the default case, the home directory is used. Recommend to set the `dir` to a fast storage device and avoid using remote filesystems, such NFS.
```

Comment (kuenishi) [reply chainerio/cache/file_cache.py]: I think the second sentence is broken, and I think the change is trivial.

**Pull request merged**